### PR TITLE
Split `lost_files_s3_db` into multiple commands 

### PR DESCRIPTION
**Pull request by i-oden**
## Before submitting this PR

1. **Description:** 
    - _From the task on Jira:_ At the moment we can run flask lost-files and see all the lost files in all the projects. However, sometimes we might want to delete lost files for one project but leave another project alone, e.g. if the project is in status In Progress and possibly still actively uploading. 
    - When I started looking through the command and tried to add the project as an option, I realised that it would become too complicated and unmaintainable. Instead I split the command into a group of 3 commands: `flask lost-files ls`, `flask lost-files add-missing-bucket` and `flask lost-files delete`.

2. **Jira task / GitHub issue:** DDS-1579
3. **How to test:** Go through the commands **one by one**, create a few projects and test that each command does what it's suppose to
4. **Type of change:** [_Check the relevant boxes in the section below_](#what-type-of-changes-does-the-pr-contain)
5. **Add docstrings and comments to code**, _even if_ you personally think it's obvious.

## What _type of change(s)_ does the PR contain?

- [x] New feature
  - [ ] Breaking: _Please describe the reason for the break and how we can fix it._
  - [x] Non-breaking
- [ ] Database change
  - [ ] Migration _included in PR_
  - [ ] Migration _not needed_
- [ ] Bug fix
  - [ ] Breaking: _Please describe the reason for the break and how we can fix it._
  - [ ] Non-breaking
- [ ] Security Alert fix
- [ ] Documentation
- [ ] Tests **(only)**
- [ ] Workflow

## Checklist

- [Sprintlog](../SPRINTLOG.md)
  - [x] Added
  - [ ] Not needed (E.g. PR contains _only_ tests)
- Rebase / Update / Merge _from_ base branch (the branch from which the current is forked)
  - [x] Done
  - [ ] Not needed
- Blocking PRs
  - [ ] Merged
  - [x] No blocking PRs
- PR to `master` branch
  - [ ] Yes: Read [the release instructions](../doc/procedures/new_release.md)
    - [ ] I have followed steps 1-7.
  - [x] No

## Actions / Scans

- **Black**: Python code formatter. Does not execute. Only tests.
  Run `black .` locally to execute formatting.
  - [x] Passed
- **Prettier**: General code formatter. Our use case: MD and yaml mainly.
  Run `npx prettier --write .` locally to execute formatting.
  - [x] Passed
- **Yamllint**: Linting of yaml files.
  - [x] Passed
- **Tests**: Pytest to verify that functionality works as expected.
  - [x] New tests added
  - [ ] No new tests
  - [ ] Passed
- **CodeQL**: Scan for security vulnerabilities, bugs, errors
  - [ ] New alerts: _Go through them and either fix, dismiss och ignore. Add reasoning in items below._
  - [ ] Alerts fixed: _What?_
  - [ ] Alerts ignored / dismissed: _Why?_
  - [x] Passed
- **Trivy**: Security scanner
  - [ ] New alerts: _Go through them and either fix, dismiss och ignore. Add reasoning in items below._
  - [ ] Alerts fixed: _What?_
  - [ ] Alerts ignored / dismissed: _Why?_
  - [x] Passed
- **Snyk**: Security scanner
  - [ ] New alerts: _Go through them and either fix, dismiss och ignore. Add reasoning in items below._
  - [ ] Alerts fixed: _What?_
  - [ ] Alerts ignored / dismissed: _Why?_
  - [x] Passed

<a href="https://gitpod.io/#https://github.com/ScilifelabDataCentre/dds_web/pull/1438"></a>

### CODE REVIEW

**Review by valyo — CHANGES_REQUESTED** [review]
Everything else looks good and works, I even tested with few thousands of files :)
```
[2023-06-16 06:52:02,865] commands [INFO] Files deleted from S3: 8703
[2023-06-16 06:52:02,866] commands [INFO] Files deleted from DB: 0
```

File: dds_web/utils.py [hunk 1]
```diff
@@ -675,3 +675,47 @@ def block_if_maintenance(user=None):
...
+    """List lost files in project."""
+    s3_filenames: set = set()
+    db_filenames: set = set()
+
+    # Get items in s3
+    try:
+        s3_filenames = set(entry.key for entry in s3_resource.Bucket(project.bucket).objects.all())
+    except s3_resource.meta.client.exceptions.NoSuchBucket:
+        missing_expected: bool = not project.is_active
+        flask.current_app.logger.info(f"Project bucket is missing. Expected: {missing_expected}")
```
[the remark this answers is not in the document]
Comment (i-oden) [reply dds_web/utils.py:691]: Will fix.

**Comment by i-oden** [comment]
@valyo I've added tests for most of it now. I was trying to add tests for when the bucket doesn't exist in `add-missing-bucket` but it was failing for unknown reason. So I'm skipping that.

You can have a look at the tests now (when you're not busy with something else)

**Review by valyo — COMMENTED** [review]
- when there are missing files `flask lost-files ls` does not return summary for the unit at the end
- `No lost files in project ...` is printed twice when project is specified:
```
/code # flask lost-files ls -p project_1
[2023-06-21 09:56:14,606] __init__ [INFO] Logging initiated.
[2023-06-21 09:56:14,779] commands [DEBUG] Searching for lost files in project 'project_1'.
[2023-06-21 09:56:14,840] utils [INFO] No lost files in project 'project_1'
[2023-06-21 09:56:14,841] commands [INFO] No lost files in project 'project_1'
```

- I still see the point of possibly listing files in DB when there is no bucket, because it seems nothing can be done anyways; I mean, even when the bucket gets created we cannot put the files found in the DB into the new bucket, we can just delete them. Or am I wrong about this?
- this inconsistency in the last message could be confusing:
```
/code # flask lost-files ls -p project_2
[2023-06-21 10:01:03,650] __init__ [INFO] Logging initiated.
[2023-06-21 10:01:03,824] commands [DEBUG] Searching for lost files in project 'project_2'.
[2023-06-21 10:01:03,879] utils [INFO] Project 'project_2' bucket is missing. Expected: False
[2023-06-21 10:01:03,880] commands [INFO] Command cancelled. Not listing files.
/code # flask lost-files ls
[2023-06-21 10:04:53,841] __init__ [INFO] Logging initiated.
[2023-06-21 10:04:54,020] commands [DEBUG] No project specified, searching for lost files in all units.
[2023-06-21 10:04:54,030] commands [INFO] Listing lost files in unit: unit_1
[2023-06-21 10:04:54,084] utils [INFO] No lost files in project 'project_1'
[2023-06-21 10:04:54,088] utils [INFO] Project 'project_2' bucket is missing. Expected: False
[2023-06-21 10:04:54,089] commands [INFO] No lost files for unit 'unit_1'
/code # flask lost-files ls -p project_1
[2023-06-21 10:05:19,116] __init__ [INFO] Logging initiated.
[2023-06-21 10:05:19,308] commands [DEBUG] Searching for lost files in project 'project_1'.
[2023-06-21 10:05:19,369] utils [INFO] No lost files in project 'project_1'
[2023-06-21 10:05:19,370] commands [INFO] No lost files in project 'project_1'
```

I am not sure how big of a deal these things are, but I am just concerned that if/when you are not around people trying to use the command(s) can get confused about what they are supposed to do. Do we have something like flowchart/diagram of the possible situations when the `lost-files` should be used and what and how can be done?

**Comment by i-oden** [comment]
> * when there are missing files `flask lost-files ls` does not return summary for the unit at the end

✅  Fixed

> * `No lost files in project ...` is printed twice when project is specified:
> 
> ```
> /code # flask lost-files ls -p project_1
> [2023-06-21 09:56:14,606] __init__ [INFO] Logging initiated.
> [2023-06-21 09:56:14,779] commands [DEBUG] Searching for lost files in project 'project_1'.
> [2023-06-21 09:56:14,840] utils [INFO] No lost files in project 'project_1'
> [2023-06-21 09:56:14,841] commands [INFO] No lost files in project 'project_1'
> ```

✅ Fixed

>     * I still see the point of possibly listing files in DB when there is no bucket, because it seems nothing can be done anyways; I mean, even when the bucket gets created we cannot put the files found in the DB into the new bucket, we can just delete them. Or am I wrong about this?

If there's a missing bucket for a project, then all files in the database will be listed as lost if we do that. Which is unneccesary. But yes, if this were to happen, we would just be able to delete them anyway, with the delete command. But there's no point in listing the files in the db if there are missing buckets.

>     * this inconsistency in the last message could be confusing:
> 
> 
> ```
> /code # flask lost-files ls -p project_2
> [2023-06-21 10:01:03,650] __init__ [INFO] Logging initiated.
> [2023-06-21 10:01:03,824] commands [DEBUG] Searching for lost files in project 'project_2'.
> [2023-06-21 10:01:03,879] utils [INFO] Project 'project_2' bucket is missing. Expected: False
> [2023-06-21 10:01:03,880] commands [INFO] Command cancelled. Not listing files.
> /code # flask lost-files ls
> [2023-06-21 10:04:53,841] __init__ [INFO] Logging initiated.
> [2023-06-21 10:04:54,020] commands [DEBUG] No project specified, searching for lost files in all units.
> [2023-06-21 10:04:54,030] commands [INFO] Listing lost files in unit: unit_1
> [2023-06-21 10:04:54,084] utils [INFO] No lost files in project 'project_1'
> [2023-06-21 10:04:54,088] utils [INFO] Project 'project_2' bucket is missing. Expected: False
> [2023-06-21 10:04:54,089] commands [INFO] No lost files for unit 'unit_1'
> /code # flask lost-files ls -p project_1
> [2023-06-21 10:05:19,116] __init__ [INFO] Logging initiated.
> [2023-06-21 10:05:19,308] commands [DEBUG] Searching for lost files in project 'project_1'.
> [2023-06-21 10:05:19,369] utils [INFO] No lost files in project 'project_1'
> [2023-06-21 10:05:19,370] commands [INFO] No lost files in project 'project_1'
> ```

I've tried to fix some inconsistency now but I'm not sure I can do more than this. 

> I am not sure how big of a deal these things are, but I am just concerned that if/when you are not around people trying to use the command(s) can get confused about what they are supposed to do. Do we have something like flowchart/diagram of the possible situations when the `lost-files` should be used and what and how can be done?

Good point. No, there's no flowchart / diagram of it, but I can write some information about this after summer.

**Comment by valyo** [comment]
>If there's a missing bucket for a project, then all files in the database will be listed as lost if we do that. Which is unneccesary. But yes, if this were to happen, we would just be able to delete them anyway, with the delete command. But there's no point in listing the files in the db if there are missing buckets.

Sorry for the typo, what I wanted to write is:
>"I still don't see the point of ..."

and it is about the message `[INFO] Command cancelled. Not listing files.`, which sounds like something's gone wrong and the command was cancelled before it could do its job; and as you wrote there's no point of listing files anyhow.

**Review by valyo — APPROVED** [review]
It works as intended and the output looks good. We could look at this again when we have some diagrams and improve it further if necessary.

**Pull request merged**